src/ch08-01-id-generation.md

```diff
@@ -1 +1,92 @@
-# Pattern: Id Generation
+# Pattern: ID Generation
+
+It should be noted that ID generation approach which was chosen in your app affects your app's architecture and overall performance.
+Here we discuss how to design good ID and show a couple of great ID designs.
+
+## Designing a great ID
+
+### Choosing the domain for your IDs
+
+Despite of wide spread of 64-bit computer architecture several popular technologies still don't support 64-bit unsigned integer numbers completely.
+For example several popular versions of Java [have only `i64` type](https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html) so you [can't have](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html#MAX_VALUE) positive integer numbers bigger than `2 ^ 63 - 1` without excess [overhead](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html).
+Therefore a lot of Java-powered apps (e.g. Graylog) have limitations on IDs' domain.
+At some external apps you can gain a couple of problems because of IDs less than `1`, e.g. value `0` may be [explicitly forbidden](https://cloud.google.com/datastore/docs/concepts/entities#assigning_your_own_numeric_id).
```

Comment:
Also avoiding `0` allows some memory optimizations ([NoneZeroU64](https://doc.rust-lang.org/std/num/struct.NonZeroU64.html)).
Besides, the use of zeros instead of Nullable columns can improve performance in some DB (e.g. ClickHouse).